### ISSUE TITLE
Undefined index when deleting array of objects

### ISSUE THREAD

**Issue by MartinP7r** [plank/laravel-mediable#217]
I'm getting this error on Laravel 8, when trying to delete all items of a relationship via

```php
$expenseReport->expenses->delete();
```

```log
ErrorException: Undefined index: App\Expense

/app/biztrip/vendor/laravel/framework/src/Illuminate/Collections/Collection.php:1367
/app/biztrip/vendor/plank/laravel-mediable/src/MediableCollection.php:74
/app/biztrip/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php:230
/app/biztrip/vendor/plank/laravel-mediable/src/MediableCollection.php:81
```

where :81 is:
```php
$classes[get_class($item)][] = $item->getKey()
```

**Comment by frasmage** [comment]
Hi @MartinP7r,

good catch! The code was doing some array access tricks on a collection which doesn't work smoothly. Fixed in 5.0.1

**Comment by MartinP7r** [comment]
Thanks a bunch. That was quick.
I just realized I was still on 4.4.2. I hope that doesn't make a difference.  
However it seems that there's a mismatch between Types there. At least it throws the below error for me:

```
TypeError: Argument 1 passed to Plank\Mediable\MediableCollection::Plank\Mediable\{closure}() must be an instance of Illuminate\Database\Eloquent\Builder, instance of Illuminate\Database\Query\Builder given

/app/biztrip/vendor/plank/laravel-mediable/src/MediableCollection.php:125
/app/biztrip/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:1464
/app/biztrip/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:699
/app/biztrip/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:843
/app/biztrip/vendor/plank/laravel-mediable/src/MediableCollection.php:128
/app/biztrip/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php:230
/app/biztrip/vendor/plank/laravel-mediable/src/MediableCollection.php:129
```

in your `MediableCollection.php`

```php
use Illuminate\Database\Eloquent\Builder; # this `Builder`
# (...)
$this->each(function (Model $item) use ($query, $relation, $classes) {
    $classes[get_class($item)][] = $item->getKey();

    $query->orWhere(function (Builder $q) use ($item, $relation) { # the `Builder` that get's passed here seems to be of type Illuminate\Database\Query\Builder 
        $q->where($relation->getMorphType(), get_class($item));
        $q->where($relation->getQualifiedForeignPivotKeyName(), $item->getKey());
    });
});
```

Please let me know if I'm seeing things wrong. Might also be a fault on my side. Maybe I'm calling `delete()` wrong...

**Comment by frasmage** [comment]
Nope, that looks like another legitimate bug. It appears that there is a missing test case around this method and it clearly hasn't been used in a while. I'll fix that this evening.

**Comment by frasmage** [comment]
Should be fixed now in 5.0.2. Thanks for bringing this one to my attention, I didn't realize it wasn't tested. I have added a shiny new integration test around the flow, to be sure it's actually working. I also optimized the delete query a little bit while I was looking at it.

**Comment by MartinP7r** [comment]
That's great! Really appreciate it!

The test is also very helpful for me as a reference for my own writing. Thanks!
